### PR TITLE
Add import and duplicate support for MCP servers

### PR DESCRIPTION
**Pull request by TravisTheTechie**
** Vibe Coded This **

- Added `import_server` method to `McpServersController` for importing server configs, with duplicate name handling.
- Added `convert_cursor_format` to convert Cursor IDE server configs to SwarmUI format.
- Implemented `duplicate` action and `McpServer#duplicate` method to allow duplicating servers with unique names.
- Improved tag handling and server type mapping.
- Added tests and factories for MCP server import and duplication.
- Required some updates because name must be unique letters, not numbers.

### CODE REVIEW

File: app/controllers/mcp_servers_controller.rb [hunk 1]
```diff
@@ -117,8 +117,17 @@ def import
       json_content = file.read
       data = JSON.parse(json_content)
 
-      # Handle both single object and array
-      servers_data = data.is_a?(Array) ? data : [data]
+      # Check configuration format
+      servers_data = if !data.is_a?(Array) && data&.key?("mcpServers")
```
Comment (TravisTheTechie) [inline app/controllers/mcp_servers_controller.rb:121]: This is great, but given these have known locations I should add a button to directly import them (might be coming next). Finding hidden files in Finder is extra steps.

File: app/models/mcp_server.rb [hunk 2]
```diff
@@ -20,9 +20,14 @@ class McpServer < ApplicationRecord
...
-  scope :with_tag, ->(tag) { where("tags @> ?", [tag].to_json) }
+  scope :with_tag, ->(tag) {
+    # SQLite doesn't support @> operator, use LIKE instead
+    where("LOWER(tags) LIKE LOWER(?)", "%\"#{tag}\"%")
+  }
   scope :search, ->(query) {
-    where("name ILIKE :q OR description ILIKE :q OR tags::text ILIKE :q", q: "%#{query}%")
+    # SQLite uses LIKE for case-insensitive searches by default (unless PRAGMA case_sensitive_like is ON)
+    # For consistent case-insensitive behavior, use LOWER()
+    where("LOWER(name) LIKE LOWER(:q) OR LOWER(description) LIKE LOWER(:q) OR LOWER(tags) LIKE LOWER(:q)", q: "%#{query}%")
```
Comment (TravisTheTechie) [inline app/models/mcp_server.rb:30]: I'm not sure if this is right, but it works. I haven't spent much time thinking about it. We control the SQLite implementation, so might be able to ignore the `LOWER` usage because we know the `PRAGMA` setting isn't updated.

File: test/controllers/mcp_servers_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,568 @@
+# frozen_string_literal: true
```
Comment (TravisTheTechie) [inline test/controllers/mcp_servers_controller_test.rb:1]: Vibed with some minor edits to get it running. It passes so I'm okay with it but I haven't done a full line-by-line on it yet.

File: test/factories/mcp_servers.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
```
Comment (TravisTheTechie) [inline test/factories/mcp_servers.rb:3]: Vibed, I'm not familiar with FactoryBot. But a little googling on the sequence I think this is working right. Also, I like this a lot better than I've done in the past (who doesn't love trying to merge kwargs). But I'll be trying to leverage this more in the future for sure.

**Review by parruda — COMMENTED** [review]
## Code Review for PR #110: Add Import and Duplicate Support for MCP Servers

Thank you for this contribution\! The import/export functionality and multi-format support are valuable additions. Here are the key issues that need attention:

### 🔴 Critical Issues

**1. Missing Model-Level Tests**
The PR lacks unit tests for the `McpServer` model. Critical validations, scopes, and the `duplicate` method are only tested indirectly through controller tests. Please add `test/models/mcp_server_test.rb` with tests for:
- Name validation rules and uniqueness
- Conditional validations for stdio/sse types
- All scopes (`with_tag`, `search`, `by_type`)
- The `duplicate` method edge cases

**2. Security Concerns with Imported Commands**
There's no validation or sanitization of imported command paths, arguments, or environment variables. This could allow:
- Path traversal attacks via command paths
- Command injection through args/env
- Malicious headers in SSE servers

Consider adding validation to ensure imported commands are safe.

**3. Database Compatibility Regression**
The SQL changes from PostgreSQL to SQLite syntax (e.g., replacing `@>` with `LIKE` for JSON) appear to drop PostgreSQL support entirely. This needs clarification:
- Is this intentional? 
- Will production still use PostgreSQL?
- The LIKE-based JSON searches are less efficient and may not scale

### 🟡 Important Issues

**4. Format Detection Fragility**
The format detection logic in lines 121-130 is brittle:
```ruby
if \!data.is_a?(Array) && data&.key?("mcpServers")  # Cursor
elsif \!data.is_a?(Array) && data["mcp"]&.key?("servers")  # VS Code
```
This will break if external formats change. Consider extracting this into a service with a registry pattern for better extensibility.

**5. Non-Intuitive Duplicate Naming**
Using letters (`"a" * counter`) for duplicate suffixes is confusing. The code mentions "validation rules" but doesn't explain WHY numbers can't be used. Please add clear documentation explaining this constraint, or consider updating the validation to allow numbers.

**6. Mixed Controller Responsibilities**
The `import` method handles too many concerns: file reading, format detection, conversion, and server creation. Consider extracting this into an `McpImportService` for better separation of concerns.

### 🟢 Good Practices Observed

- Comprehensive controller test coverage
- Good use of factories for test data
- Proper handling of multiple import formats
- Clean test organization with proper teardown

### Recommendations

1. **Extract format conversion logic** into service objects for better testability and maintainability
2. **Add security validations** for imported configurations
3. **Clarify database strategy** and ensure compatibility with production PostgreSQL
4. **Document the letter-based naming constraint** clearly in code comments
5. **Add missing model and security tests** before merging

The functionality is solid, but these issues should be addressed to ensure security, maintainability, and production readiness.

**Comment by TravisTheTechie** [comment]
## Updated Review: Significant Improvements Made 🎉

Thank you for addressing the feedback\! The refactored code with service-oriented architecture is a **major improvement**. Here's my updated assessment:

### ✅ Critical Issues Successfully Fixed (4/5)

1. **Nil value crash** ✅ FIXED - Proper `.is_a?(Hash)` checks prevent crashes
2. **VS Code SSE data loss** ✅ FIXED - Correctly detects type based on fields
3. **File size limit** ✅ FIXED - 10MB limit enforced
4. **Name sanitization** ✅ FIXED - Comprehensive handling of all edge cases
5. **Duplicate naming** ⚠️ PARTIAL - Still uses letters (a,b,c) instead of numbers

### 🏗️ Architecture Improvements

✅ **Service-oriented design implemented** 
- Clean separation with `McpServerImporter` and format-specific classes
- Controller reduced from 317 to 196 lines (38% reduction)
- Follows Rails conventions and existing patterns
- Easy to extend for new formats

### 📊 Overall Assessment

| Aspect | Status | Score |
|--------|--------|-------|
| Code Correctness | 4/5 issues fixed | ✅ Good |
| Architecture | Clean service pattern | ✅ Excellent |
| Maintainability | Much improved | ✅ Good |
| Test Coverage | Controller tests updated | ⚠️ Adequate |

### 🔧 Minor Remaining Issues

1. **Duplicate naming pattern** - Still uses confusing letters (copy_a, copy_b) instead of intuitive numbers (copy_2, copy_3)
2. **No service class tests** - Services only tested indirectly through controller tests
3. **Missing documentation** - No class/method documentation for the new services

### Recommendation: APPROVE ✅

The PR has addressed the critical blocking issues and implemented a clean architecture. The remaining issues are minor and can be addressed in follow-up PRs:

- Consider switching to numeric suffixes for better UX
- Add service class unit tests for better maintainability
- Add documentation for the service interfaces

Great work on the refactoring\! The service-oriented approach makes this much more maintainable and extensible.
